Add OrderedDictionary

stephentoub opened the pull request:
Closes https://github.com/dotnet/runtime/issues/24826

stephentoub committed:
Add OrderedDictionary

kronic commented:
@stephentoub @terrajobst 
Maybe it's worth adding more methods
```csharp
public void Move(int oldIndex, int newIndex);
public void Stort(IComparer<TValue> comparer);
```

eiriktsarpalis commented:
Adding some kind of sorting functionality would be useful, but ideally that should be solved for IList<T> in general, cf. https://github.com/dotnet/runtime/issues/76375

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1188 @@
+        where TKey : notnull
+    {
+        /// <summary>Store for the key/value pairs in the dictionary.</summary>
+        private readonly Dictionary<TKey, TValue> _dictionary;
```
The approach followed by STJ in this case is using a `List<KeyValuePair<TKey, TValue>>` as the primary store with linear lookup and only materializing a dictionary if the size exceeds a fixed threshold:

https://github.com/dotnet/runtime/blob/f690a1a16611449fe281fe3ba4065939986d95be/src/libraries/System.Text.Json/src/System/Text/Json/JsonPropertyDictionary.cs#L16-L19

This should avoid the need to make additional lookups while enumerating (which presumably   is an important operation in ordered dictionaries).

stephentoub replied:
That helps the case where there are < 9 entries and hurts the case where there are >= 9 entries. We believe < 9 is going to be the majority case? I can understand that being common for JSON, but I'm not convinced that translates to here.

eiriktsarpalis replied:
It is true that it would incur one additional check. We can always measure impact on perf for both cases.

> We believe < 9 is going to be the majority case?

Perhaps not the majority, but I'm sure it would impact a substantial number of instances. The fact that we're picking a starting default initial capacity of 3 reflects an expectation that enough of those instances won't grow beyond that threshold.

I don't particularly mind if we always instantiate a dictionary though. More important is whether we want to use a `List<KeyValuePair<TKey, TValue>>` and how that impacts enumeration performance.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1188 @@
+        /// <summary>Store for the key/value pairs in the dictionary.</summary>
+        private readonly Dictionary<TKey, TValue> _dictionary;
+        /// <summary>List storing the keys in order.</summary>
+        private readonly List<TKey> _list;
```
This could be replaced with a growable array.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1188 @@
+
+                if (value is not KeyValuePair<TKey, TValue> tpair)
+                {
+                    throw new ArgumentException(SR.Format(SR.Arg_WrongType, value, typeof(KeyValuePair<TKey, TValue>)), nameof(value));
```
Do we care that this throw hasn't been factored into a stub method?

stephentoub replied:
I don't think so, until perf tests prove otherwise. Other collections in System.Collections have open-coded throws rather than using throw helpers for everything.

TylerBrinkley commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1188 @@
+        where TKey : notnull
+    {
+        /// <summary>Store for the key/value pairs in the dictionary.</summary>
+        private readonly Dictionary<TKey, TValue> _dictionary;
```
Is there a reason we're not using a dedicated data store like in my existing implementation? https://github.com/dotnet/corefxlab/blob/archive/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs

stephentoub replied:
Because it adds a lot of complexity, especially to mirror all the optimizations in `Dictionary<>`, e.g. that existing implementation doesn't special-case strings. If we want to optimize it later based on demonstrated need that it's worth the complexity, we can.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1188 @@
+        /// </summary>
+        public OrderedDictionary()
+        {
+            _dictionary = [];
```
This constructor will default to whatever initial capacity is being used by each of the underlying collections. Should we make sure that this is set uniformly?

stephentoub replied:
We _could_, but maintaining that would be challenging, given that they have different growth strategies (list doubles, dictionary doubles and rounds up to the next prime). Is there a functional concern with keeping them the exact same? If we really want that fine-grained level of control, we can go the other route and open-code the whole implementation rather than relying on a pair of Dictionary/List, I just didn't think it was worth the non-trivial amount of code duplication and complexity right now to do so... do you think otherwise?

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1188 @@
+            {
+                if (_keyEnumerator.MoveNext())
+                {
+                    Current = new(_keyEnumerator.Current, _dictionary._dictionary[_keyEnumerator.Current]);
```
In particular, I find this bit to be problematic. Enumerating through the dictionary results in a full lookup of every key that we know is present in the collection. I'm pretty sure that would materialize as a performance regression if I were to use this in STJ property serialization.

alexrp replied:
I'm also a bit concerned about what the performance would end up looking like here as I do quite a lot of enumeration over the key-value pairs. I currently use @TylerBrinkley's implementation [which just indexes an array](https://github.com/dotnet/corefxlab/blob/e0c8705741ad2b66a1bbc70ff5bf5e02886eca1e/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs#L918).

stephentoub replied:
Ok. If we believe enumeration is really the top consideration, I'll redo it.

stephentoub replied:
>> If we're really concerned about this, we should scrap using a Dictionary/List separately and do the much more complex but feasible thing of open-coding the whole data structure.

> I don't think we have to go that far.

What else would you do? If enumeration is really the top consideration, then its perf shouldn't fall off a cliff after 9 elements.

eiriktsarpalis replied:
Whether we make the dictionary optional for small counts is orthogonal to the enumeration concern.

```C#
private readonly Dictionary<TKey, int> _dictionary;
private readonly List<KeyValuePair<TKey, TValue>> _entries;
```

But if we wanted to make it optional this would result in `_dictionary` being created lazily and only being used in key-based lookup. For enumeration the list would always remain the source of truth.

stephentoub replied:
> But if we wanted to make it optional this would result in _dictionary being created lazily and only being used in key-based lookup.

Yes, but then inserting / removing requires a lookup for each key in order to update the index; Dictionary doesn't expose anything that lets you update a value for a key as part of an enumeration.

stephentoub replied:
> Good point. I guess duplicating the value in both underlying stores is another trade-off?

If the dictionary's value is `(int, TValue)`, then you still have O(N) lookups on remove / insert.

If the dictionary's value is just `TValue`, then `this[key] = value` is an O(N) operation to keep the list's copy in sync.

eiriktsarpalis replied:
It appears that @TylerBrinkley's implementation also incurs O(n) cost for update operations. So it seems we have to make a choice between optimizing for reads and optimizing for writes?

stephentoub replied:
> also incurs O(n) cost for update operations.

So does this one, since `List<T>` needs to do a memory shift. It's not a question of the algorithmic complexity but of the constant factor; are you doing an optimized memcpy, or are you walking O(N) elements but doing a simple write per, or are you doing O(N) full lookups on a hash table.

> So it seems we have to make a choice between optimizing for reads and optimizing for writes?

Yes

TylerBrinkley replied:
> It appears that @TylerBrinkley's implementation also incurs O(n) cost for update operations.

Unless I'm mistaken I believe my implementation is O(n) only for `Insert` and `Remove` unless resizing is needed. See the chart here.

https://github.com/dotnet/runtime/issues/24826#issuecomment-362044463

stephentoub committed:
Open-code collection data structure rather than using Dictionary+List

This rewrites the core of the type to be based on a custom data structure rather than wrapping Dictionary and List. The core structure is based on both Dictionary in corelib and the OrderedDictionary prototype in corefxlab.

This also adds missing TryAdd, Capacity, EnsureCapacity, and TrimExcess members for parity with Dictionary, and fixes debugger views for the Key/ValueCollections.

stephentoub commented:
@eiriktsarpalis, please take another look when you get a chance. I redid the guts based on the feedback to prioritize enumeration. It's based on a combination of approaches taken by `Dictionary<>` and by the corefxlab `OrderedDictionary<>`.

stephentoub commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+            }
+
+            /// <inheritdoc/>
+            TKey IList<TKey>.this[int index]
```
Thanks, but I think it's still logically ambiguous to have an indexer on the collection and an indexer on collection.Keys/Values that end up meaning different things.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+                _comparer = comparer ?? EqualityComparer<TKey>.Default;
+
+                if (typeof(TKey) == typeof(string) &&
+                    NonRandomizedStringEqualityComparer.GetStringComparer(_comparer!) is IEqualityComparer<string> stringComparer)
```
Is the suppression necessary here?
```suggestion
                    NonRandomizedStringEqualityComparer.GetStringComparer(_comparer) is IEqualityComparer<string> stringComparer)
```

stephentoub replied:
It shouldn't be. I'll double check.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+            //   devirtualize and possibly inline the operation.
+            if (!typeof(TKey).IsValueType)
+            {
+                _comparer = comparer ?? EqualityComparer<TKey>.Default;
```
Nit: this logic is reading and writing the `_comparer` field in a number of places, could this be simplified by calculating the `comparer` and then storing it to the field?

stephentoub replied:
We explicitly want to leave the field null for value types (if no custom comparer was provided) so that later on we use `EqualityComparer<T>.Default.Equals` to get devirtualization and inlining. This is copy/paste what Dictionary does.

eiriktsarpalis replied:
The local is nullable, so it could be nulled out if necessary. Keeping consistency with Dictionary is fine though.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+            {
+                ArgumentNullException.ThrowIfNull(key);
+
+                bool modified = TryInsert(-1, key, value, InsertionBehavior.OverwriteExisting);
```
```suggestion
                bool modified = TryInsert(index: -1, key, value, InsertionBehavior.OverwriteExisting);
```

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+                        break;
+
+                    default:
+                        return false;
```
What's the insertion behavior handled by the default case? A Debug assert might better highlight this.

stephentoub replied:
TryAdd, where nothing happens if the key already exists.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+                switch (behavior)
+                {
+                    case InsertionBehavior.OverwriteExisting:
+                        _entries[i].Value = value;
```
Do we ignore the `index` parameter in this case?

stephentoub replied:
Yes. We could assert its < 0 if that's desirable; it always will be.

eiriktsarpalis replied:
I would guess it's predicated on callers of this method to never pass a non-negative index combined with overwrite behavior. Asserting for that precondition couldn't hurt.

stephentoub replied:
Yes

stephentoub committed:
Address PR feedback

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+            /// <summary>A snapshot of the dictionary's version when enumeration began.</summary>
+            private readonly int _version;
+            /// <summary>Whether Current should be a DictionaryEntry.</summary>
+            private readonly bool _useDictionaryEntry;
```
Is this hybrid mode used  by the dictionary enumerator as well?

stephentoub replied:
Yup

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1864 @@
+                {
+                    Debug.Assert(dictionary._entries is not null);
+                    ref Entry entry = ref dictionary._entries[_index];
+                    Current = new KeyValuePair<TKey, TValue>(entry.Key, entry.Value);
```
Nit: looks like the ctor could be target typed?

stephentoub replied:
We've generally avoided doing that when the type isn't visible on the lhs.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/ThrowHelper.cs:
```diff
@@ -0,0 +1,36 @@
+
+        /// <summary>Throws an exception for an index being out of range.</summary>
+        [DoesNotReturn]
+        internal static void ThrowIndexOutOfRange() =>
```
```suggestion
        internal static void ThrowArgumentOutOfRange() =>
```

eiriktsarpalis commented on src/libraries/System.Collections/tests/Generic/OrderedDictionary/OrderedDictionary.Generic.Tests.Keys.cs:
```diff
@@ -0,0 +1,105 @@
+
+        protected override ICollection<string> GenericICollectionFactory(int count)
+        {
+            OrderedDictionary<string, string> dictionary = new OrderedDictionary<string, string>();
```
```suggestion
            OrderedDictionary<string, string> dictionary = new();
```

stephentoub committed:
Add more tests based on code coverage gaps

stephentoub committed:
Address PR feedback

stephentoub committed:
Merge remote-tracking branch 'upstream/main' into ordereddictionary

stephentoub committed:
Try to fix NativeAOT tests in CI

stephentoub commented:
/ba-g all failures are known

AndreyAkinshin commented:
@stephentoub Thank you very much for doing this! It's fantastic that we can finally use generic OrderedDirectiory out of the box without using 3rd-party NuGet packages or implementing it from scratch.

eiriktsarpalis commented on src/libraries/System.Collections/src/System/Collections/Generic/OrderedDictionary.cs:
```diff
@@ -0,0 +1,1867 @@
+            // to record the newly updated indices.
+            for (i = _count - 1; i >= index; --i)
+            {
+                entries[i + 1] = entries[i];
```
Is there a chance we could perf improvement if the shifting of the entries array used block copying?

stephentoub replied:
Possibly. But we still need to walk the entries individually to fix up the chains. We can more easily experiment with it once it's in a daily.
